runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1200,44 +1432,60 @@ static iree_status_t iree_hal_hip_device_queue_execute(
     iree_hal_command_buffer_t* command_buffer,
     iree_hal_buffer_binding_table_t binding_table) {
   iree_hal_hip_device_t* device = iree_hal_hip_device_cast(base_device);
-
   IREE_TRACE_ZONE_BEGIN(z0);
+
+  if (queue_affinity == IREE_HAL_QUEUE_AFFINITY_ANY) {
+    queue_affinity = 0x1;
+  }
+
+  uint64_t queue_affinity_mask =
+      ((iree_hal_queue_affinity_t)1 << device->topology.count);
+  queue_affinity_mask = queue_affinity_mask | (queue_affinity_mask - 1);
+  queue_affinity &= queue_affinity_mask;
+
+  uint64_t nidx = iree_math_count_trailing_zeros_u64(queue_affinity);
+  queue_affinity = (uint64_t)1 << nidx;
+
+  iree_hal_hip_device_semaphore_submit_callback_data_t* callback_data;
   IREE_RETURN_AND_END_ZONE_IF_ERROR(
-      z0, iree_hal_hip_set_context(device->hip_symbols, device->hip_context));
+      z0, iree_hal_hip_device_make_callback_data(
+              device, device->host_allocator, &device->block_pool,
+              queue_affinity, wait_semaphore_list, signal_semaphore_list,
+              command_buffer, binding_table, &callback_data));
+  iree_status_t status = iree_ok_status();
 
-  iree_status_t status = iree_hal_deferred_work_queue_enqueue(
-      device->work_queue, iree_hal_hip_device_collect_tracing_context,
-      device->tracing_context, wait_semaphore_list, signal_semaphore_list,
-      command_buffer ? 1 : 0, command_buffer ? &command_buffer : NULL,
-      &binding_table);
-  if (iree_status_is_ok(status)) {
-    // Try to advance the deferred work queue.
-    status = iree_hal_deferred_work_queue_issue(device->work_queue);
+  if (wait_semaphore_list.count == 0) {
```

Comment:
same as above, use vertical whitespace to group blocks of logic
```suggestion

  iree_status_t status = iree_ok_status();
  if (wait_semaphore_list.count == 0) {
```